File: 2016/s1.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using std::cin;
using std::cout;
using std::endl;
using std::string;

#define map std::unordered_map

/** Gets the letters of string and puts their occurences in a dictionary */
map<char, int> getLetters(string str) {
    map<char, int> letters;

    for (char& letter : str) {
        letters[letter]++;
    }

    return letters;
}

int main() {
    string str1;
    string str2;

    cin >> str1;
    cin >> str2;

    auto str1Letters = getLetters(str1);
    auto str2Letters = getLetters(str2);
    bool didFinish = false;

    for (auto const& entry : str1Letters) {
        char letter = entry.first;
        int str1Count = entry.second;
        int str2Count = str2Letters[letter];
        int letterCountDifference = str1Count - str2Count;

        if (letterCountDifference == 0) {
            str2Letters[letter] = 0;
        } else if (letterCountDifference <= -1) {
            cout << "N" << endl;
            didFinish = true;

            break;
        } else {
            int wildCardCount = str2Letters['*'];

            if (wildCardCount >= letterCountDifference) {
                str2Letters['*'] = wildCardCount - letterCountDifference;
                str2Letters[letter] = 0;
            } else {
                cout << "N" << endl;
                didFinish = true;

                break;
            }
        }
    }

    if (!didFinish) {
        cout << "A";
    }

    return 0;
}

File: 2016/s1.py
#!/bin/python

from typing import Dict

string1 = input()
string2 = input()


def getLetters(string: str) -> Dict[str, int]:
    """Gets the letters of string and puts their occurences in a dictionary"""
    letters: Dict[str, int] = {}

    for letter in string:
        letters[letter] = letters.get(letter, 0) + 1

    return letters


string1Letters = getLetters(string1)
string2Letters = getLetters(string2)

for letter, string1Count in string1Letters.items():
    string2Count = string2Letters.get(letter, 0)

    letterCountDifference = string1Count - string2Count

    # If letter counts are equal
    if letterCountDifference == 0:
        string2Letters[letter] = 0
    # If string2 has more of a letter than string1, break
    elif letterCountDifference <= -1:
        print("N")

        break
    # Try to use wildcards
    else:
        wildCardCount = string2Letters.get("*", 0)

        # If wildcards available for use
        if wildCardCount >= letterCountDifference:
            string2Letters["*"] = wildCardCount - letterCountDifference
            string2Letters[letter] = 0
        else:
            print("N")

            break
else:
    print("A")
